addons/popochiu/engine/objects/character/popochiu_character.gd

```diff
@@ -742,6 +756,10 @@ func face_direction(destination: Vector2):
 	# Get the vector from the origin to the destination.
 
 	var angle = wrapf(rad_to_deg((destination - position).angle()), 0, 360)
+	var direction_faced = int(angle/45)
```

Comment:
You know? I'm absolutely not sure that a division, plus and integer casting, plus a variable assignment, plus an if branch, aren't slower than 16 comparisons in a worse case.

Did you benchmark it, somehow?

Reply:
I didn't benchmark it, so possibly 16 comparisons are actually faster. 
When I was talking about speed I was only referring to doing the 16 comparisons as a binary tree to limit that code section to max 4 comparisons.
This PR isn't primarily about speed (though it should be faster than the original as a lot less code is called). 
The origin of the PR was that I had to step through the code to trace something, and it was calling so many functions and evaluating so many code lines each frame that didn't need to keep getting called when values weren't changing.

Reply:
OK, maybe the best option is to avoid going through a long `if/else` chain altogether.
I'm sorry for not elaborating more on this, but I have very little time at the moment, but... What if we avoid testing the angle against the values and calculate its "direction index" something like this:

```gdscript
	var angle = wrapf(rad_to_deg((destination - position).angle()), 0, 360)

	# Normalize the angle to a 0-7 range index (8 directions)
	var direction_index := int((angle + 22.5) / 45.0) % 8
	
	# Map direction index to Looking enum and animation suffixes
	match direction_index:
		0: # Right (337.5-22.5)
			_looking_dir = Looking.RIGHT
			_animation_suffixes = ['_r', '_l', '_dr', '_dl', '_d']
		1: # Down-Right (22.5-67.5) 
			_looking_dir = Looking.DOWN_RIGHT
			_animation_suffixes = ['_dr', '_dl', '_d', '_r', '_l']
		2: # Down (67.5-112.5)
			_looking_dir = Looking.DOWN
			_animation_suffixes = ['_d', '_dr', '_dl', '_r', '_l'] 
		3: # Down-Left (112.5-157.5)
			_looking_dir = Looking.DOWN_LEFT
			_animation_suffixes = ['_dl', '_dr', '_l', '_r', '_d']
		4: # Left (157.5-202.5)
			_looking_dir = Looking.LEFT
			_animation_suffixes = ['_l', '_r', '_dl', '_dr', '_d']
		5: # Up-Left (202.5-247.5)
			_looking_dir = Looking.UP_LEFT
			_animation_suffixes = ['_ul', '_ur', '_u', '_l', '_r']
		6: # Up (247.5-292.5)
			_looking_dir = Looking.UP
			_animation_suffixes = ['_u', '_ul', '_ur', '_l', '_r']
		7: # Up-Right (292.5-337.5)
			_looking_dir = Looking.UP_RIGHT
			_animation_suffixes = ['_ur', '_ul', '_r', '_l', '_u']
```

This code seems to work from a very quick test. I hope I didn't mistaken the math, but if it's ok, we can do a further step forward and get rid of the match using a map.

What I mean is, instead of "matching" the index, we can arrange the "Looking" enum so that the 0 value is RIGHT.
Also, we can have an array map for the direction suffixes, which keys are the corresponding values of the enum, and we can do:

```gdscript
func face_direction(destination: Vector2):
    var angle = wrapf(rad_to_deg((destination - position).angle()), 0, 360)
    # this requires the Looking enum to be ordered from RIGHT, clockwise
    _looking_dir := int((angle + 22.5) / 45.0) % 8
    # this requires an array or dictionary available to store the suffixes
    _animation_suffixes = _valid_animation_suffixes[_looking_dir] + ['']
```

If you want to avoid the last assignment, the _looking_dir value can be checked against a cache, but since you would have to test and reassign the _looking_dir value to the cache anyway, it's probably better to leave the `_animation_suffixes` assignment, maybe.

What do you think @balloonpopper? Makes sense?

Reply:
I'd originally thought of doing a look up table but went with my original idea because I thought that while it's fairly code efficient, it's not the most readable and easy to debug if the maths isn't quite right. I can change it to a look up table though.

Reply:
I don't think it's so hard to read. We can always leave comments with the calculated range and our reasoning in functions. It can feel like cluttering the code, but it makes a difference for ourselves and new contributors.

I think a table is the way to go: efficient, quick, and easy to maintain.

Reply:
Modified to use a table. I thought I had worked out why you added 22.5 degrees, then when looking at the code later, I couldn't remember. I tried without it and it works fine.